book/src/dev/rfcs/0003-state-updates.md

```diff
@@ -0,0 +1,305 @@
+heights to blocks. Chains are stored using an ordered map from difficulty to
+chains, so that the map ordering is the ordering of best to worst chains.
+
+- XXX fill in details on exact types
+
+- XXX work out whether we should store extra data (e.g., a HashSet of UTXOs
+  spent by some block etc.) to speed up checks.
+
+When a new block extends the best chain past 100 blocks, the old root is
+removed from the in-memory state and committed to sled.
+
+## Sled data structures
+[sled]: #sled
+
+Sled provides a persistent, thread-safe `BTreeMap<&[u8], &[u8]>`. Each map is
+a distinct "tree". Keys are sorted using lex order on byte strings, so
+integer values should be stored using big-endian encoding (so that the lex
+order on byte strings is the numeric ordering).
+
+We use the following Sled trees:
+
+| Tree                 |                  Keys |                              Values |
+|----------------------|-----------------------|-------------------------------------|
+| `blocks_by_hash`     | `BlockHeaderHash`     | `Block`                             |
+| `hash_by_height`     | `BE32(height)`        | `BlockHeaderHash`                   |
+| `tx_by_hash`         | `TransactionHash`     | `BlockHeaderHash || BE32(tx_index)` |
+| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
+| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
+| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
+
+Zcash structures are encoded using `ZcashSerialize`/`ZcashDeserialize`.
+
+## Request / Response API
+[request-response]: #request-response
+
+The state API is provided by a pair of `Request`/`Response` enums. Each
+`Request` variant corresponds to particular `Response` variants, and it's
+fine (and encouraged) for caller code to unwrap the expected variants with
+`unreachable!` on the unexpected variants. This is slightly inconvenient but
+it means that we have a unified state interface with unified backpressure.
+
+This API includes both write and read calls. Spotting `Commit` requests in
+code review should not be a problem, but in the future, if we need to
+restrict access to write calls, we could implement a wrapper service that
+rejects these, and export "read" and "write" frontends to the same inner service.
+
+### `Request::CommitBlock(Arc<Block>)`
+[request-commit-block]: #request-commit-block
+
+Performs contextual validation of the given block, committing it to the state
+if successful. Returns `Response::Added(BlockHeaderHash)` with the hash of
+the newly committed block or an error.
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.
+
+Otherwise, attempt to perform contextual validation checks and the commit the
+given block to the state. The exact list of contextual validation checks will
+be specified in a later RFC. If contextual validation checks succeed, the new
```

Comment:
> attempt to perform contextual validation checks and the commit the
given block to the state. The exact list of contextual validation checks will
be specified in a later RFC.

What do these contextual validation checks look like, and where do they happen?

There are two conflicting design principles here, and I'd like to explore the tradeoff:
* related functions should be performed together, which suggests that all high-level block checks should be executed by the `BlockVerifier` service
* related data should be managed in the same place, which suggests that contextual validation should happen in `zebra-state`

We have a few different options here:
1. Move `BlockVerifier` to `zebra_state`, and add a synchronous (non-async) function which performs contextual validation on a `Block`. This function is invoked from the state service's `call` method
2. Split block checks between `zebra_consensus::BlockVerifier`, which is async, and `zebra_state::ContextVerifier`, which is synchronous. `zebra_state::ContextVerifier`'s functions are invoked from the state service's `call` method.

I think option 1 makes it easier to add checks in each network upgrade, because the relevant verification code is all in the same place. It might make it easier to review existing checks, and it might make the data dependencies clearer.

I think option 2 is more modular, and it has a clearer separation of concerns. But I don't like the way it splits verification and data dependencies across multiple crates.

Let's chat about the options here, when everyone is available?

Reply:
I think options 1 or 2 make it easier to add checks in each network upgrade, because the relevant verification code is all in the same place. They might make it easier to review existing checks, and it might make the data dependencies clearer.

I think options 2 or 3 are more modular, and they have a clearer separation of concerns.

But I don't like the way option 3 splits verification and data dependencies across multiple crates.

What was the intention of the RFC here?

Reply:
I think that (3) is the current plan, and I think that it's useful to have the split because it segments the checks into the stateful and stateless ones.  It definitely splits the verification checks, but I don't think that it splits data dependencies across multiple crates, because all of the contextual checks are isolated to the state system.

Reply:
Seems like a good plan!